fpdf/text_region.py

```diff
@@ -0,0 +1,275 @@
+        current_y = self.pdf.y
+        for paragraph in self._paragraphs:
+            cur_lines = paragraph._build_lines(current_y, print_sh)
+            if not cur_lines:
+                continue
+            current_y = paragraph.current_y
+            text_lines.extend(cur_lines)
+        return text_lines
+
+    def render(self, print_sh: bool = False):
+        if not self._paragraphs:
+            return False
+        text_lines = self.collect_lines(print_sh)
+        return self._render_lines(text_lines)
+
+    def get_width(self, height):
+        limits = self.current_x_extents(self.pdf.y, height)
+        res = limits[1] - max(self.pdf.x, limits[0]) - 2 * self.pdf.c_margin
+        return res
+
+
+class TextColumnarMixin:
+    """Enable a TextRegion to perform page breaks"""
+
+    def __init__(self, pdf, *args, l_margin=None, r_margin=None, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.l_margin = pdf.l_margin if l_margin is None else l_margin
+        left = self.l_margin
+        self.r_margin = pdf.r_margin if r_margin is None else r_margin
+        right = pdf.w - self.r_margin
+        self._set_left_right(left, right)
+
+    def _set_left_right(self, left, right):
+        self.left = self.pdf.l_margin if left is None else left
+        self.right = (self.pdf.w - self.pdf.r_margin) if right is None else right
+        if self.right <= self.left:
+            raise FPDFException(
+                f"{self.__class__.__name__}(): "
+                f"Right limit ({self.right}) lower than left limit ({self.left})."
+            )
+
+    def current_x_extents(self, y, height):  # pylint: disable=unused-argument
+        """Return the horizontal extents of the current line.
+        Columnar regions simply return the boundaries of the column.
+        Regions with non-vertical boundaries need to check how the largest
+        font-height in the current line actually fits in there.
+        For that reason we include the current y and the line height.
+        """
+        return self.left, self.right
+
+
+class TextColumns(TextRegion, TextColumnarMixin):
+    def __init__(self, pdf, *args, ncols: int = 1, gap_width: float = 10, **kwargs):
+        super().__init__(pdf, *args, **kwargs)
+        self.cur_column = 0
+        self.cur_top = self.pdf.t_margin
+        self.ncols = ncols
+        self.gap_width = gap_width
+        total_w = self.right - self.left
+        self.col_width = (total_w - (self.ncols - 1) * self.gap_width) / self.ncols
+        # We calculate the column extents once in advance, and store them for lookup.
+        # This way we can later also enable the possibility to request columns of
+        # differing width.
+        c_left = self.left
+        self.cols = [(c_left, c_left + self.col_width)]
+        for i in range(1, ncols):  # pylint: disable=unused-variable
+            c_left += self.col_width + self.gap_width
+            self.cols.append((c_left, c_left + self.col_width))
```

Comment:
For clarity, `self.cols` should maybe contain `NamedTuples` with `.start` & `.end` attributes?

Reply:
For use in one single class?

Reply:
Well, you are manipulating "extents" values all over this module 😊

Reply:
Maybe it will make sense once there are some shaped region classes.